BUG: Correct lapack ld* args

eric-wieser opened the pull request:
This is split off from #3861

eric-wieser committed:
BUG: Work around bugs in LAPACK 3.0.0

eric-wieser committed:
BUG: Correctly compute the LDA parameters for lapack calls

These are documented in the lapack source code as always being one or greater.
LD stands for "leading dimension", and can be thought of as a stride.

eric-wieser commented on numpy/linalg/umath_linalg.c.src:
```diff
@@ -706,6 +706,11 @@ dump_@TYPE@_matrix(const char* name,
+#ifndef min
+#define min(x,y) ((x<y)?x:y)
+#define max(x,y) ((x>y)?x:y)
+#endif
```
Copied from `:/numpy/random/mtrand/distributions.c`

eric-wieser commented on numpy/linalg/umath_linalg.c.src:
```diff
@@ -1250,51 +1257,57 @@ init_@lapack_func@(EIGH_PARAMS_t* params, char JOBZ, char UPLO,
+        liwork = query_iwork_size;
+    }
 
-    work = mem_buff;
```
This line was just wrong before

seberg commented:
Sorry, not finding a lot of time.... First glance, looks fine to me, but I really don't know linalg well at all. Eric, how does this fit now with the remove axis stuff? Would this be merged first, and then the other one can avoid some of the hackier stuff?

eric-wieser commented:
Yep, ideally this would be merged first, and then #3861 would probably not have to touch linalg at all

eric-wieser commented:
@seberg: Do you want me to go ahead and rebase that on this?

charris commented on numpy/linalg/umath_linalg.c.src:
```diff
@@ -706,6 +706,11 @@ dump_@TYPE@_matrix(const char* name,
  */
 
+#ifndef min
+#define min(x,y) ((x<y)?x:y)
```
shouldn't max have a guard also?

charris replied:
And spaces around operators. Macros don't give you a pass here ;)

charris replied:
Also, macros are in CAPS.

eric-wieser replied:
Thought that I could get a pass here by virtue of copying `/numpy/random/mtrand/distributions.c
`. Should I fix the macros there too?

charris replied:
Probably. There are lots of small things like that scattered through existing code that need fixing but don't have a high priority. For new code we try to get it right...

eric-wieser replied:
Wouldn't "right" look like an inline function then, to make `max(a++, b++)` work correctly?

eric-wieser replied:
@charris: How does the latest commit look to address this? If it's alright, I'll squash it over the previous commit

eric-wieser commented on numpy/linalg/umath_linalg.c.src:
```diff
@@ -706,10 +706,21 @@ dump_@TYPE@_matrix(const char* name,
-#define min(x,y) ((x<y)?x:y)
-#define max(x,y) ((x>y)?x:y)
-#endif
+static inline fortran_int
```
Somewhat surprisingly, this file doesn't use `NPY_INLINE` anywhere

charris replied:
Hmm, that is strange. Then again, the npy version was needed when not all compilers supported the C99 standard and that may no longer be a problem.

charris commented on numpy/linalg/umath_linalg.c.src:
```diff
@@ -1244,57 +1262,75 @@ typedef struct eigh_params_struct {
+
+    /* Work size query */
+    {
+        @typ@ query_work_size;
```
Any reason not to keep these declaration up top and remove the block? I think it would read better...

eric-wieser replied:
We already used [a block further down the file for this](https://github.com/numpy/numpy/pull/8756/files#diff-a7b89b86bed09822472a1300f5f9feefL2496), so one reason is just consistency.

But I also think it's more clear this way it makes it obvious which things are only used as temporaries for the work size.

charris commented on numpy/linalg/umath_linalg.c.src:
```diff
@@ -1306,29 +1342,30 @@ init_@lapack_func@(EIGH_PARAMS_t* params, char JOBZ, char UPLO,
+/**end repeat**/
+
 
+/**begin repeat
```
Spaces around `=` would look better.

charris replied:
And after `,`.

eric-wieser replied:
Wasn't sure that was legal. Will do

eric-wieser replied:
Actually, these are like this because they're unmodified, and git has decided to produce a diff to suggest they have been. The actual change was just to move init_@lapack_func@ after call_@lapack_func@, but one of these functions looks very much like another

Do you want me to add a STY commit adding spacing on top of everything else?

charris commented on numpy/linalg/umath_linalg.c.src:
```diff
@@ -1358,39 +1392,50 @@ init_@lapack_func@(EIGH_PARAMS_t *params,
-    lrwork = (fortran_int)query_rwork_size;
-    liwork = query_iwork_size;
+    /* Work size query */
+    {
```
Again, any need for this block?

charris commented on numpy/linalg/umath_linalg.c.src:
```diff
@@ -1358,39 +1392,50 @@ init_@lapack_func@(EIGH_PARAMS_t *params,
+        params->RWORK = &query_rwork_size;
+        params->IWORK = &query_iwork_size;
+
+        if (call_@lapack_func@(params) != 0)
```
All if entries should be blocks
```
if (...) {
    single line;
}
```

eric-wieser replied:
Even this is an inherited style violation

charris commented on numpy/linalg/umath_linalg.c.src:
```diff
@@ -1358,39 +1392,50 @@ init_@lapack_func@(EIGH_PARAMS_t *params,
     mem_buff2 = malloc(lwork*sizeof(@typ@) +
                        lrwork*sizeof(@basetyp@) +
                        liwork*sizeof(fortran_int));
     if (!mem_buff2)
```
See above on `if`

charris commented:
There are a ton of (inherited) style violations. May be best to put those fixes off to another PR. I am not a fan of freestanding blocks, although they do have their uses, but I don't see why they are needed here.

eric-wieser reviewed (commented):
Did you mean to close this?

charris commented:
No

eric-wieser commented:
Do you want the style one merged first or second?

charris commented:
We have the following to deal with Microsoft's version of C.
```
#if defined(_MSC_VER)
        #define NPY_INLINE __inline
#elif defined(__GNUC__)
	#if defined(__STRICT_ANSI__)
		#define NPY_INLINE __inline__
	#else
		#define NPY_INLINE inline
	#endif
#else
        #define NPY_INLINE
#endif
```

Not sure why `inline` worked before.

charris commented:
Let's make it the second one unless you have a preference.

eric-wieser commented:
In that case, is the only remaining objection here the blocks?

eric-wieser committed:
MAINT: Move definitions of call_ before init_

This diff makes it look like something else has happened, but in reality
that's the only change here.

This means we can invoke call_@lapack_func@ in init_@lapack_func@ when we do
workspace size calculations

eric-wieser committed:
MAINT: Always use parameter structs to call lapack

Getting the arguments to lapack calls in the correct order in two places is
harder than doing it in one.

eric-wieser commented:
I've rebased to try and isolate the broken diff from moving functions around

charris commented:
Let's leave the blocks to the style fixup also. It will probably be easier to see.

eric-wieser commented:
> Let's leave the blocks to the style fixup also.

Leave removing them, as in accept this as is? Or debate adding them in the style fixup?

charris commented:
I cannot figure why the `inline` modifier is working on windows. Or are we compiling this on windows?

eric-wieser commented:
I am definitely compiling this on windows, and it works just fine

eric-wieser commented:
@charris: All green

eric-wieser commented:
> I cannot figure why the inline modifier is working on windows. Or are we compiling this on windows?I cannot figure why the inline modifier is working on windows. Or are we compiling this on windows?

Because `/numpy/core/setup.py` contains

```python
            # define inline to our keyword, or nothing
            target_f.write('#ifndef __cplusplus\n')
            if inline == 'inline':
                target_f.write('/* #undef inline */\n')
            else:
                target_f.write('#define inline %s\n' % inline)
            target_f.write('#endif\n')
```

charris commented:
But linalg isn't *in* numpy/core...

eric-wieser commented:
Thanks for reviewing, @charris

eric-wieser commented:
> But linalg isn't in numpy/core...

True, but we `#include "npy_config.h"`, which _is_ in `numpy/core`
